venv issue with make manager

## Basics

Have you updated your local code to the latest available from the repo: yes

Have you ran `make update` after updating your local code: no

What version of HMS-Docker are you currently using (run `cat /opt/hms-docker/.hmsd-version`), or is this a new install: new install

## Operating System and Version

```bash
cat /etc/os-release
ansible --version
```


OS name and version: 
"Ubuntu 24.04.4 LTS"
fresh install

## Describe the bug

A clear and concise description of what the bug is.
I followed the instructions: https://hmsdocker.dev/docs/getting-started/install/

## Expected behavior

i expected make manager to run normally after i installed the suggested python3-venv package.

## List any applicable variables and their values

user@plex:~/ansible-hms-docker$ make config
⚠ This will overwrite all existing files in 'inventory/group_vars/all', are you sure? [y/n] > y
[OK] Copying files...
user@plex:~/ansible-hms-docker$ make manager
Creating Python virtual environment
The virtual environment was not created successfully because ensurepip is not
available.  On Debian/Ubuntu systems, you need to install the python3-venv
package using the following command.

    apt install python3.12-venv

You may need to use sudo with that command.  After installing the python3-venv
package, recreate your virtual environment.

Failing command: /home/user/ansible-hms-docker/.venv/bin/python3

Installing Python requirements
/bin/bash: line 5: .venv/bin/pip3: No such file or directory
make: *** [Makefile:83: manager] Error 127
user@plex:~/ansible-hms-docker$ sudo apt install python3.12-venv
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following additional packages will be installed:
  python3-pip-whl python3-setuptools-whl
The following NEW packages will be installed:
  python3-pip-whl python3-setuptools-whl python3.12-venv
0 upgraded, 3 newly installed, 0 to remove and 0 not upgraded.
Need to get 2429 kB of archives.
After this operation, 2777 kB of additional disk space will be used.
Do you want to continue? [Y/n] 
Get:1 http://nl.archive.ubuntu.com/ubuntu noble-updates/universe amd64 python3-pip-whl all 24.0+dfsg-1ubuntu1.3 [1707 kB]
Get:2 http://nl.archive.ubuntu.com/ubuntu noble-updates/universe amd64 python3-setuptools-whl all 68.1.2-2ubuntu1.2 [716 kB]
Get:3 http://nl.archive.ubuntu.com/ubuntu noble-updates/universe amd64 python3.12-venv amd64 3.12.3-1ubuntu0.15 [5674 B]
Fetched 2429 kB in 0s (10.6 MB/s)                              
debconf: delaying package configuration, since apt-utils is not installed
Selecting previously unselected package python3-pip-whl.
(Reading database ... 103920 files and directories currently installed.)
Preparing to unpack .../python3-pip-whl_24.0+dfsg-1ubuntu1.3_all.deb ...
Unpacking python3-pip-whl (24.0+dfsg-1ubuntu1.3) ...
Selecting previously unselected package python3-setuptools-whl.
Preparing to unpack .../python3-setuptools-whl_68.1.2-2ubuntu1.2_all.deb ...
Unpacking python3-setuptools-whl (68.1.2-2ubuntu1.2) ...
Selecting previously unselected package python3.12-venv.
Preparing to unpack .../python3.12-venv_3.12.3-1ubuntu0.15_amd64.deb ...
Unpacking python3.12-venv (3.12.3-1ubuntu0.15) ...
Setting up python3-setuptools-whl (68.1.2-2ubuntu1.2) ...
Setting up python3-pip-whl (24.0+dfsg-1ubuntu1.3) ...
Setting up python3.12-venv (3.12.3-1ubuntu0.15) ...
Scanning processes...                                                                                                                                                                                                                                                                                                                                            
Scanning processor microcode...                                                                                                                                                                                                                                                                                                                                  
Scanning linux images...                                                                                                                                                                                                                                                                                                                                         

Running kernel seems to be up-to-date.

The processor microcode seems to be up-to-date.

No services need to be restarted.

No containers need to be restarted.

No user sessions are running outdated binaries.

No VM guests are running outdated hypervisor (qemu) binaries on this host.
user@plex:~/ansible-hms-docker$ make manager
Traceback (most recent call last):
  File "/home/user/ansible-hms-docker/settings_manager.py", line 4, in <module>
    from ruamel.yaml import YAML
ModuleNotFoundError: No module named 'ruamel'
make: *** [Makefile:84: manager] Error 1
user@plex:~/ansible-hms-docker$ rm -rf .venv
user@plex:~/ansible-hms-docker$ make manager
Creating Python virtual environment
Installing Python requirements
🔧 YAML Configuration Editor

? Select an option: 📄 plex

--- a/Makefile
+++ b/Makefile
@@ -11,6 +11,11 @@ CUSTOM_CONF_DIR = inventory/group_vars/all
 
 REMOTE_CONFIG_URL = https://hmsdocker.dev/configs
 
+VENV_DIR = .venv
+VENV_PY = $(VENV_DIR)/bin/python3
+# Marker file used to detect when requirements.txt changed since the last install
+VENV_STAMP = $(VENV_DIR)/.requirements-installed
+
 ARCH = $(shell uname -m)
 BIN_DIR = ./bin
 YQ_LOCAL = $(BIN_DIR)/yq
@@ -80,13 +85,32 @@ verify-containers:
 	@sudo python3 .github/workflows/scripts/check_containers.py
 
 manager:
-	@if [ ! -d ".venv" ]; then \
+	@set -e; \
+	if [ -d "$(VENV_DIR)" ] && ! "$(VENV_PY)" -m pip --version > /dev/null 2>&1; then \
+		printf $(_WARN) "WARN" "Found an incomplete virtual environment in '$(VENV_DIR)', recreating it"; \
+		rm -rf "$(VENV_DIR)"; \
+	fi; \
+	if [ ! -d "$(VENV_DIR)" ]; then \
 		echo "Creating Python virtual environment"; \
-		python3 -m venv .venv; \
+		if ! python3 -m venv "$(VENV_DIR)"; then \
+			rm -rf "$(VENV_DIR)"; \
+			pyver=$$(python3 -c 'import sys; print("%d.%d" % sys.version_info[:2])' 2>/dev/null); \
+			printf $(_ERROR) "FAIL" "Could not create the Python virtual environment in '$(VENV_DIR)'"; \
+			printf $(_ERROR) "FAIL" "On Debian/Ubuntu, install the venv module and re-run 'make manager':"; \
+			printf $(_ERROR) "FAIL" "  sudo apt install -y python3-venv python$$pyver-venv"; \
+			exit 1; \
+		fi; \
+	fi; \
+	if [ ! -f "$(VENV_STAMP)" ] || [ requirements.txt -nt "$(VENV_STAMP)" ]; then \
 		echo "Installing Python requirements"; \
-		.venv/bin/pip3 install -r requirements.txt > /dev/null; \
-	fi;
-	@.venv/bin/python3 settings_manager.py
+		if ! "$(VENV_DIR)/bin/pip3" install -r requirements.txt > "$(VENV_DIR)/pip-install.log" 2>&1; then \
+			cat "$(VENV_DIR)/pip-install.log"; \
+			printf $(_ERROR) "FAIL" "Failed to install Python requirements from requirements.txt"; \
+			exit 1; \
+		fi; \
+		touch "$(VENV_STAMP)"; \
+	fi
+	@$(VENV_PY) settings_manager.py
 
 update: $(YQ_LOCAL)
 	@echo "Version Before: $$($(YQ_LOCAL) '.[0].vars.hmsd_current_version' hms-docker.yml)"

--- a/Makefile
+++ b/Makefile
@@ -87,6 +87,11 @@ verify-containers:
 manager:
 	@set -e; \
 	if [ -d "$(VENV_DIR)" ] && ! "$(VENV_PY)" -m pip --version > /dev/null 2>&1; then \
+		if [ ! -w "$(VENV_DIR)" ]; then \
+			printf $(_ERROR) "FAIL" "'$(VENV_DIR)' is incomplete and not writable by $$(id -un), remove it and re-run 'make manager':"; \
+			printf $(_ERROR) "FAIL" "  sudo rm -rf $(VENV_DIR)"; \
+			exit 1; \
+		fi; \
 		printf $(_WARN) "WARN" "Found an incomplete virtual environment in '$(VENV_DIR)', recreating it"; \
 		rm -rf "$(VENV_DIR)"; \
 	fi; \
@@ -94,21 +99,32 @@ manager:
 		echo "Creating Python virtual environment"; \
 		if ! python3 -m venv "$(VENV_DIR)"; then \
 			rm -rf "$(VENV_DIR)"; \
-			pyver=$$(python3 -c 'import sys; print("%d.%d" % sys.version_info[:2])' 2>/dev/null); \
+			pyver=$$(python3 -c 'import sys; print("%d.%d" % sys.version_info[:2])' 2>/dev/null) || pyver=""; \
 			printf $(_ERROR) "FAIL" "Could not create the Python virtual environment in '$(VENV_DIR)'"; \
 			printf $(_ERROR) "FAIL" "On Debian/Ubuntu, install the venv module and re-run 'make manager':"; \
-			printf $(_ERROR) "FAIL" "  sudo apt install -y python3-venv python$$pyver-venv"; \
+			if [ -n "$$pyver" ]; then \
+				printf $(_ERROR) "FAIL" "  sudo apt install -y python3-venv python$$pyver-venv"; \
+			else \
+				printf $(_ERROR) "FAIL" "  sudo apt install -y python3 python3-venv"; \
+			fi; \
 			exit 1; \
 		fi; \
 	fi; \
-	if [ ! -f "$(VENV_STAMP)" ] || [ requirements.txt -nt "$(VENV_STAMP)" ]; then \
+	if [ ! -w "$(VENV_DIR)" ]; then \
+		printf $(_WARN) "WARN" "'$(VENV_DIR)' is not writable by $$(id -un), skipping the requirements check"; \
+		printf $(_WARN) "WARN" "If the settings manager fails to start, run 'sudo rm -rf $(VENV_DIR)' and try again"; \
+	elif [ ! -f "$(VENV_STAMP)" ] || [ requirements.txt -nt "$(VENV_STAMP)" ]; then \
 		echo "Installing Python requirements"; \
-		if ! "$(VENV_DIR)/bin/pip3" install -r requirements.txt > "$(VENV_DIR)/pip-install.log" 2>&1; then \
-			cat "$(VENV_DIR)/pip-install.log"; \
+		piplog=$$(mktemp); \
+		if ! "$(VENV_PY)" -m pip install -r requirements.txt > "$$piplog" 2>&1; then \
+			cat "$$piplog"; \
+			rm -f "$$piplog"; \
 			printf $(_ERROR) "FAIL" "Failed to install Python requirements from requirements.txt"; \
 			exit 1; \
 		fi; \
-		touch "$(VENV_STAMP)"; \
+		rm -f "$$piplog"; \
+		touch "$(VENV_STAMP)" 2>/dev/null || \
+			printf $(_WARN) "WARN" "Could not write to '$(VENV_DIR)', requirements will be re-checked on every run"; \
 	fi
 	@$(VENV_PY) settings_manager.py
 

--- a/docs-astro/src/content/docs/docs/getting-started/install.mdx
+++ b/docs-astro/src/content/docs/docs/getting-started/install.mdx
@@ -39,9 +39,7 @@ This playbook configures Docker to use the `172.27.0.0/16` IP range for all Dock
 
     :::note
 
-    `python3-venv` is required by `make manager`. On some Ubuntu/Debian releases the generic
-    `python3-venv` package does not pull in the module for your interpreter, in which case also
-    install the versioned package (e.g. `sudo apt install python3.12-venv -y`).
+    `python3-venv` is required by `make manager`.
 
     :::
 

--- a/docs-astro/src/content/docs/docs/getting-started/install.mdx
+++ b/docs-astro/src/content/docs/docs/getting-started/install.mdx
@@ -34,8 +34,16 @@ This playbook configures Docker to use the `172.27.0.0/16` IP range for all Dock
 2. Install requirements:
 
     ```bash
-    sudo apt install git make python3-pip -y
+    sudo apt install git make python3-pip python3-venv -y
     ```
+
+    :::note
+
+    `python3-venv` is required by `make manager`. On some Ubuntu/Debian releases the generic
+    `python3-venv` package does not pull in the module for your interpreter, in which case also
+    install the versioned package (e.g. `sudo apt install python3.12-venv -y`).
+
+    :::
 
 3. Clone the repository:
 
@@ -75,6 +83,10 @@ There is a script that will allow you to easily edit all variable files mentione
 ```bash
 make manager
 ```
+
+This creates a Python virtual environment in `.venv/` on first run and installs the requirements
+into it. If the environment is missing or incomplete (for example, it was created before
+`python3-venv` was installed), it is rebuilt automatically.
 
 :::
 

